runtime/lua/nvim/spellfile.lua

```diff
@@ -0,0 +1,278 @@
+  local ok = (not err) and ((status >= 200 and status < 300) or (status == 0 and file_ok(outpath)))
+  return ok, (status ~= 0 and status or nil), err
+end
+
+---@return string[]
+function M.directory_choices()
+  local opts = {}
+  for _, dir in ipairs(rtp_list()) do
+    local spelldir = vim.fs.joinpath(vim.fn.fnamemodify(dir, ':p'), 'spell')
+    if vim.fn.isdirectory(spelldir) == 1 then
+      table.insert(opts, spelldir)
+    end
+  end
+  return opts
+end
+
+function M.choose_directory()
+  local dirs = writable_spell_dirs_from_rtp()
+  if #dirs == 0 then
+    return ensure_target_dir()
+  elseif #dirs == 1 then
+    return dirs[1]
+  end
+  local prompt ---@type string[]
+  prompt = {}
+  for i, d in
+    ipairs(dirs --[[@as string[] ]])
+  do
+    prompt[i] = string.format('%d: %s', i, d)
+  end
+  local choice = vim.fn.inputlist(prompt)
+  if choice < 1 or choice > #dirs then
+    return dirs[1]
+  end
+  return dirs[choice]
+end
+
+function M.parse(lang)
+  local code = normalize_lang(lang)
+  local enc = 'utf-8'
+  local dir = ensure_target_dir()
+
+  local missing = {}
+  local candidates = {
+    string.format('%s.%s.spl', code, enc),
+    string.format('%s.%s.sug', code, enc),
+  }
+  for _, fn in ipairs(candidates) do
+    if not file_ok(vim.fs.joinpath(dir, fn)) then
+      table.insert(missing, fn)
+    end
+  end
+
+  return {
+    files = missing,
+    key = code .. '.' .. enc,
+    lang = code,
+    encoding = enc,
+    dir = dir,
+  }
+end
+
+---@param info SpellInfo
+function M.download(info)
+  local dir = info.dir or ensure_target_dir()
+  if not dir then
+    notify('No (writable) spell directory found and could not create one.', vim.log.levels.ERROR)
+    return
+  end
+
+  local lang = info.lang
+  local enc = info.encoding
+
+  local spl_utf8 = string.format('%s.%s.spl', lang, enc)
+  local spl_ascii = string.format('%s.ascii.spl', lang)
+  local sug_name = string.format('%s.%s.sug', lang, enc)
+
+  local url_utf8 = M.config.url .. '/' .. spl_utf8
+  local out_utf8 = vim.fs.joinpath(dir, spl_utf8)
+  notify('Downloading ' .. spl_utf8 .. ' …')
+  local ok, st, err = http_get_to_file_sync(url_utf8, out_utf8, M.config.timeout_ms)
+  if not ok then
+    notify(
+      ('Could not get %s (status %s): trying %s …'):format(
+        spl_utf8,
+        tostring(st or 'nil'),
+        spl_ascii
+      )
+    )
+    local url_ascii = M.config.url .. '/' .. spl_ascii
+    local out_ascii = vim.fs.joinpath(dir, spl_ascii)
+    local ok2, st2, err2 = http_get_to_file_sync(url_ascii, out_ascii, M.config.timeout_ms)
+    if not ok2 then
+      notify(
+        ('No spell file available for %s (utf8:%s ascii:%s) — %s'):format(
+          lang,
+          tostring(st or err or 'fail'),
+          tostring(st2 or err2 or 'fail'),
+          url_utf8
+        ),
+        vim.log.levels.WARN
+      )
+      vim.schedule(function()
+        vim.cmd('echo ""')
+      end)
+      M._done[info.key] = true
```

Comment:
Seems like an alternative to `M._done` would be to accept callbacks in `download`:

    download(..., on_done, on_err)